tools/traffic_vault_migrate/postgres.go

```diff
@@ -0,0 +1,743 @@
+	stride := 2
+	queryArgs := make([]interface{}, len(tbl.Records)*stride)
+	for i, record := range tbl.Records {
+		j := i * stride
+		queryArgs[j] = record.DeliveryService
+		queryArgs[j+1] = record.DataEncrypted
+	}
+	return insertIntoTable(db, queryBase, stride, queryArgs)
+}
+func (tbl *pgURLSigKeyTable) gatherKeys(db *sql.DB) error {
+	sz, err := getSize(db, "url_sig_key")
+	if err != nil {
+		log.Println("PGURLSigKey gatherKeys: unable to determine url_sig_key table size")
+	}
+	tbl.Records = make([]pgURLSigKeyRecord, sz)
+
+	rows, err := db.Query("SELECT deliveryservice, data from url_sig_key")
+	if err != nil {
+		return fmt.Errorf("PGURLSigKey gatherKeys error creating query: %w", err)
+	}
+	defer rows.Close()
+	i := 0
+	for rows.Next() {
+		if i > len(tbl.Records)-1 {
+			return fmt.Errorf("PGURLSigKey gatherKeys: got more results than expected %v", len(tbl.Records))
+		}
+		err := rows.Scan(&tbl.Records[i].DeliveryService, &tbl.Records[i].DataEncrypted)
+		if err != nil {
+			return fmt.Errorf("PGURLSigKey gatherKeys: unable to scan row: %w", err)
+		}
+		i += 1
+	}
+	return nil
+}
+func (tbl *pgURLSigKeyTable) decrypt(aesKey []byte) error {
+	for i, sig := range tbl.Records {
+		err := decryptInto(aesKey, sig.DataEncrypted, &tbl.Records[i].Keys)
+		if err != nil {
+			return fmt.Errorf("unable to decrypt into keys: %w", err)
+		}
+	}
+	return nil
+}
+func (tbl *pgURLSigKeyTable) encrypt(aesKey []byte) error {
+	for i, sig := range tbl.Records {
+		data, err := json.Marshal(&sig.Keys)
+		if err != nil {
+			return fmt.Errorf("encrypt issue marshalling keys: %w", err)
+		}
+
+		dat, err := encrypt(data, aesKey)
+		if err != nil {
+			return fmt.Errorf("encrypt error: %w", err)
+		}
+		tbl.Records[i].DataEncrypted = dat
+	}
+	return nil
+}
+func (tbl *pgURLSigKeyTable) toGeneric() []URLSigKey {
+	keys := make([]URLSigKey, len(tbl.Records))
+
+	for i, record := range tbl.Records {
+		keys[i] = URLSigKey{
+			DeliveryService: record.DeliveryService,
+			URLSigKeys:      record.Keys,
+		}
+	}
+	return keys
+}
+func (tbl *pgURLSigKeyTable) fromGeneric(keys []URLSigKey) {
+	tbl.Records = make([]pgURLSigKeyRecord, len(keys))
+
+	for i, key := range keys {
+		tbl.Records[i] = pgURLSigKeyRecord{
+			Keys:            key.URLSigKeys,
+			DeliveryService: key.DeliveryService,
+			pgCommonRecord: pgCommonRecord{
+				DataEncrypted: nil,
+			},
+		}
+	}
+}
+func (tbl *pgURLSigKeyTable) validate() []string {
+	for i, record := range tbl.Records {
+		if record.DataEncrypted == nil && len(record.Keys) > 0 {
+			return []string{fmt.Sprintf("URl Sig Key %v: DataEncrypted is blank!", i)}
+		}
+	}
+	return nil
+}
+
+type pgURISignKeyRecord struct {
+	Keys            map[string]tc.URISignerKeyset
+	DeliveryService string
+	pgCommonRecord
+}
+type pgURISignKeyTable struct {
+	Records []pgURISignKeyRecord
+}
+
+func (tbl *pgURISignKeyTable) insertKeys(db *sql.DB) error {
+	queryBase := "INSERT INTO uri_signing_key (deliveryservice, data) VALUES "
```

Comment:
Expensive in terms of dev time or runtime? And do you mean skip it by `ON CONFLICT IGNORE` or something like that?

If we're talking expensive as in runtime, doing a migration should be pretty rare. If we're successful, we might only use this tool once. If we aren't successful at first, maybe we'll run it 3 times if we decide to un-migrate then re-migrate once the issue is fixed. So if it takes a little longer to run, that's probably not a big deal.

Reply:
`ON CONFLICT` doesn't work because the only constrains on the sslkey table are on the `id` column. Regardless, I'll handle it for sslkeys as well.

Reply:
Oh, I see what you're saying. We could currently have multiple sslkeys at the same version in the DB, because `version` is not part of the primary key?

Reply:
Yep, at least from the DBs perspective; I don't think we can actually do that through the TO API.

Reply:
Well, we can still change the schema if you'd like -- since it's unreleased we don't necessarily need a migration and can just update the create_schema.sql file.

Reply:
That would be preferable, makes it so I don't have to use a CTE. I'll add it to this PR (otherwise testing the psql backend would be impossible).

Reply:
I also ended up removing the `id` column as it appears that the only use for it was for the unique constraint.